stdlib/public/Observation/Sources/Observation/ObservedChanges.swift

```diff
@@ -0,0 +1,58 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+import _Concurrency
+
+@available(SwiftStdlib 5.9, *)
+public struct ObservedChanges<Subject: Observable, Element: Sendable> {
+  let context: ObservationRegistrar<Subject>.Context
+  let keyPath: KeyPath<Subject, Element>
+  
+  init(
+    _ context: ObservationRegistrar<Subject>.Context, 
+    keyPath: KeyPath<Subject, Element>
+  ) {
+    self.context = context
+    self.keyPath = keyPath
+  }
+}
+
+@available(SwiftStdlib 5.9, *)
+extension ObservedChanges: AsyncSequence {
+  public struct Iterator: AsyncIteratorProtocol {
+    let context: ObservationRegistrar<Subject>.Context
+    let keyPath: KeyPath<Subject, Element>
+    let keyPaths: KeyPaths<Subject>
+    
+    init(
+      _ context: ObservationRegistrar<Subject>.Context, 
+      keyPath: KeyPath<Subject, Element>
+    ) {
+      self.context = context
+      self.keyPath = keyPath
+      keyPaths = Subject.dependencies(of: keyPath)
+    }
+    
+    public mutating func next() async -> Element? {
+      await context.nextChange(to: keyPath, keyPaths: keyPaths)
+    }
+  }
+  
+  public func makeAsyncIterator() -> Iterator {
+    Iterator(context, keyPath: keyPath)
+  }
+}
+
+@available(SwiftStdlib 5.9, *)
+extension ObservedChanges: @unchecked Sendable where Subject: Sendable { }
+
+@available(*, unavailable)
+extension ObservedChanges.Iterator: Sendable { }
```

Comment:
Why is the `Sendable` conformance okay for the sequence, but not the iterator? It looks like the context is going to mutated out from under both.

Reply:
`AsyncIterator`s must not be `Sendable`. We learned this the hard way. The problem with them being `Sendable` is that you cannot tie a specific consumer task to an iterator anymore. This makes it super hard to implement the `AsyncSequence` itself because it has to suddenly handle multiple consumers on a single iterator even though it is a unicast `AsyncSequence`. Our latest stance on this is that `AsyncSequence`s should most of the time be `Sendable` and can be passed to different tasks. Each task should create its own iterator to consume the sequence.

If an `AsyncSequence` doesn't support multiple iterators we are `fatalError`ing on this in most places right now.

Reply:
the AsyncSequence definition is just a description of what to observe which means that if the subject in question can be sent across tasks the the changes associated with said observation can also be sent across tasks. However the iteration infers stateful cancellation and mutation of the context of which next element is being iterated, that cannot be sent meaningfully because if one task cancels then all tasks would get nil, and more of an issue is that if one task consumes a next value that value is denied to other tasks.

So the sequence itself is sendable but the iterator is not.

Reply:
Interesting! This all sounds very similar to the headaches around copying a regular sequence iterator. (If we designed things from scratch today, perhaps iterators would not be required to be copyable.)

Reply:
Yeah, that's my reaction — it's surprising that you can make multiple copies of the iterator and then iterate them concurrently, drawing from the same underlying `context`, but the `Sendable` semantics don't explicitly ban such a system. The cancellation bit is what I was missing in the difference between the sequence(s) and iterator(s). Thanks!